src/ImageSharp/Formats/Tiff/Writers/TiffBiColorWriter.cs

```diff
@@ -3,208 +3,97 @@
-    /// </summary>
-    internal class TiffBiColorWriter : TiffBaseColorWriter
+    internal class TiffBiColorWriter<TPixel> : TiffBaseColorWriter<TPixel>
+        where TPixel : unmanaged, IPixel<TPixel>
     {
-        public TiffBiColorWriter(TiffStreamWriter output, MemoryAllocator memoryAllocator, Configuration configuration, TiffEncoderEntriesCollector entriesCollector)
-            : base(output, memoryAllocator, configuration, entriesCollector)
-        {
-        }
+        private readonly Image<TPixel> imageBlackWhite;
 
-        /// <summary>
-        /// Writes the image data as 1 bit black and white to the stream.
-        /// </summary>
-        /// <typeparam name="TPixel">The pixel data.</typeparam>
-        /// <param name="image">The image to write to the stream.</param>
-        /// <param name="quantizer">The quantizer.</param>
-        /// <param name="compression">The compression to use.</param>
-        /// <param name="compressionLevel">The compression level for deflate compression.</param>
-        /// <param name="useHorizontalPredictor">if set to <c>true</c> [use horizontal predictor].</param>
-        /// <returns>
-        /// The number of bytes written.
-        /// </returns>
-        public override int Write<TPixel>(Image<TPixel> image, IQuantizer quantizer, TiffEncoderCompression compression, DeflateCompressionLevel compressionLevel, bool useHorizontalPredictor)
-        {
-            int padding = image.Width % 8 == 0 ? 0 : 1;
-            int bytesPerRow = (image.Width / 8) + padding;
-            using IMemoryOwner<L8> pixelRowAsGray = this.MemoryAllocator.Allocate<L8>(image.Width);
-            using IManagedByteBuffer row = this.MemoryAllocator.AllocateManagedByteBuffer(bytesPerRow, AllocationOptions.Clean);
-            Span<byte> outputRow = row.GetSpan();
-            Span<L8> pixelRowAsGraySpan = pixelRowAsGray.GetSpan();
+        private IMemoryOwner<byte> pixelsAsGray;
+
+        private IMemoryOwner<byte> bitStrip;
 
+        public TiffBiColorWriter(ImageFrame<TPixel> image, MemoryAllocator memoryAllocator, Configuration configuration, TiffEncoderEntriesCollector entriesCollector)
+            : base(image, memoryAllocator, configuration, entriesCollector)
+        {
             // Convert image to black and white.
             // TODO: Should we allow to skip this by the user, if its known to be black and white already?
-            using Image<TPixel> imageBlackWhite = image.Clone();
-            imageBlackWhite.Mutate(img => img.BinaryDither(default(ErrorDither)));
-
-            if (compression == TiffEncoderCompression.Deflate)
-            {
-                return this.WriteBiColorDeflate(imageBlackWhite, pixelRowAsGraySpan, outputRow, compressionLevel);
-            }
+            this.imageBlackWhite = new Image<TPixel>(configuration, new ImageMetadata(), new[] { image.Clone() });
+            this.imageBlackWhite.Mutate(img => img.BinaryDither(default(ErrorDither)));
```

Comment:
`BiColor` encoder use hard `Mutate` operation, I think this is the reason for the slow results of benchmarks

Reply:
we can check the source image, if it is already `bicolor`, we can use source directly

Reply:
`default(ErrorDither)` is a big red flag here, it won't produce any output since there's no matrix. We should be using one of the static properties. `ErrorDither.FloydSteinberg` or similar. Looks like LibTiff uses that one. http://www.libtiff.org/man/tiffmedian.1.html

>we can check the source image, if it is already bicolor, we can use source directly

We cannot do that since it will almost always be different from the output.

Reply:
> `default(ErrorDither)` is a big red flag here, it won't produce any output since there's no matrix. We should be using one of the static properties. `ErrorDither.FloydSteinberg` or similar. Looks like LibTiff uses that one. http://www.libtiff.org/man/tiffmedian.1.html

@JimBobSquarePants yeah, it's required to use `ErrorDither.FloydSteinberg` (or another that we can add to encoder options ), in this PR I didn't change this instruction, I just moved this instruction to a new method. 

> > we can check the source image, if it is already bicolor, we can use source directly
> 
> We cannot do that since it will almost always be different from the output.
Yeah, it's a narrow cases, for example, when re-saving or joining several monochrome scans into single tiff (in the future).
I mean, if we had `L1` (`L8` analogy) we can skipping hard dithering operation.

Reply:
> @JimBobSquarePants yeah, it's required to use ErrorDither.FloydSteinberg (or another that we can add to encoder options > ?), in this PR I didn't change this instruction, I just moved this instruction to a new method.

@IldarKhayrutdinov this was my mistake, i can change that in the tiff branch, you dont have to do that (but that might result in a merge conflict here)

Reply:
@brianpopow Yeah, there are still such mistakes, and mine too 😄 but we're close to RC!
So that there are no conflicts, it is better on make in this PR, one of these days I will deal with remarks

Reply:
Do you think it could be the cause of your blank images?

Reply:
> Do you think it could be the cause of your blank images?

shouldn't be out of this, on weekend I'll debugging empty images

Reply:
> Do you think it could be the cause of your blank images?

Im still trying to figure out the reason for the black images. At the moment i think its not actually in the encoder, but rather an issue with the decoder.
If you just save the image like this:

```
image.Save("output.tiff", new TiffEncoder()
{
    Compression = TiffEncoderCompression.Lzw,
    Mode = TiffEncodingMode.ColorPalette
});
```

The image looks fine and `tiffinfo` from libtiff does not complain about it.

Reply:
@brianpopow could you attach your output?

Reply:
here is an output example with the above mentioned settings:

[lzw_colorpalette.zip](https://github.com/SixLabors/ImageSharp/files/5975930/lzw_colorpalette.zip)

Im working currently on an replacement implementation for the lzw decoder. Looks good so far but i dont want to promise too much until its done. Im pretty sure the black image issue is the decoder.

Reply:
The black image issue is now fixed in the `tiff-format` branch.